standard/ranges.md

```diff
@@ -0,0 +1,309 @@
+
+## 24.1 General
+
+This clause introduces a model for *extended indexable* and *sliceable* *collection* types built on:
+
+- The types introduced in this clause, `System.Index` (§24.2) and `System.Range` (§24.3);
+- The pre-defined unary `^` (§hat-operator) and binary `..` (§range-operator) operators; and
+- The *element_access* expression.
+
+Under the model a type is classified as:
+
+- a *collection* if it represents a group of *element*s
+- an *extended indexable* collection if it supports an *element_access* expression which has a single argument expression of type `Index` which returns and/or sets a single element of the type, either by value or by reference; and
+- an *extended sliceable* collection if it supports an *element_access* expression which has a single argument expression of type `Range` which returns a *slice* of the elements of the type by value.
+
+> *Note*: The model does not require that a slice, unlike an element, of the type can be set, but a type may support it as an extension of the model. *end note*
+
+The model is supported for single-dimensional arrays (§12.8.12.2) and strings (§string-access).
+
+The model can be supported by any class, struct or interface type which provides appropriate indexers (§15.9) which implement the model semantics.
+
+Implicit support for the model is provided for types which do not directly support it but which provide a certain *pattern* of members (§24.4). This support is pattern-based, rather than semantic-based, as the semantics of the type members upon which it is based are *assumed* – the language does not enforce, or check, the semantics of these type members.
+
+For the purposes of this clause the following terms are defined:
+
+- A ***collection*** is a type which represents a group of ***element***s.
+- A ***countable collection*** is one which provides a ***countable property*** an `int` valued instance property whose value is the number of elements currently in the group. This property shall be named either `Length` or `Count`, the former is chosen if both exist.
+- A ***sequence*** or ***indexable type*** is a collection:
+  - which is countable;
+  - where every element can be accessed using an *element_access* expression with a single required `int` argument, the ***from-start index***, additional optional arguments are allowed;
+  - a sequence is ***modifiable*** if every element can also be set using an *element_access* expression;
+  - an element’s from-start index is the number of elements before it in the sequence, for a sequence containing *N* elements:
+    - the first and last elements have indices of 0 and *N*-1 respectively, and
+    - the ***past-end index***, an index which represents a hypothetical element after the last one, has the value *N*.
+- A ***from-end index*** represents an element’s position within a sequence relative to the last element. For a sequence containing *N* elements the first, last and past-end indices are *N*, 1 and 0 respectively.
```

Comment:
`^1` refers to the last element, so the `1` is not a position relative to the last element.
```suggestion
- A ***from-end index*** represents an element’s position within a sequence relative to the past-end index. For a sequence containing *N* elements the first, last and past-end indices are *N*, 1 and 0 respectively.
```

Reply:
Agreed